File: C++/543-Diameter_of_Binary_Tree.cpp
// Runtime: 12 ms, faster than 69.34% of C++ online submissions for Diameter of Binary Tree.
// Memory Usage: 19.7 MB, less than 92.59% of C++ online submissions for Diameter of Binary Tree.
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int diameterOfBinaryTree(TreeNode* root) {
        int res = 0;
        dfs(root, res);
        return res;
    }
private:
    int dfs(TreeNode* node, int& res) {
        if (!node)
            return 0;
        int left = dfs(node->left, res);
        int right = dfs(node->right, res);
        res = max(res, left + right);
        return max(left, right) + 1;
    }
};

File: Algorithms/Python3.x/543-Diameter_of_Binary_Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def diameterOfBinaryTree(self, root: 'TreeNode') -> 'int':
        self.n = 1
        self.depth(root)
        return self.n - 1

    def depth(self, node):
        if not node: return 0
        left = self.depth(node.left)
        right = self.depth(node.right)
        self.n = max(self.n, left + right + 1)
        return max(left, right) + 1
